module: Added flawfinder checker for finding security flaws

leemgs opened the pull request:
Fixed issue #522.

This module is to examines C/C++ source code and reports possible security weaknesses.

Flawfinder  searches  through C/C++ source code looking for potential security flaws. It examines
all of the project's C/C++ source code. It's very useful for quickly finding and removing some
security problems before a program is widely released.

Flawfinder produces a list of ``hits'' (potential security flaws), sorted by risk; the riskiest
hits are shown first.  The risk level is shown inside square brackets and  varies  from  0,
very little risk, to 5, great risk.  This risk level depends not only on the function,
but on the values of the parameters of the function.  For example, constant strings are often less risky
than fully variable strings in many contexts, and in those contexts the hit will have a lower risk level.
Hit descriptions also note the relevant Common Weakness Enumeration (CWE) identifier(s) in parentheses.

**Reference**
 * Wiki page for this module: https://github.com/nnsuite/TAOS-CI/wiki/TAOS-CI-Plug-in-Modules
 * Module guide: https://github.com/nnsuite/TAOS-CI/blob/master/ci/doc/how-to-use-taos-ci-module.md

**Changelog**
* Version 2:
  - Added annotation for the module maintenance.
  - Improved the code readability.

* Version 1:
  - Added new module (e.g. flawfinder) for scanning security flaws.
  - Added configuration variables.
  - Added document.
  - Added installation step of "flawfinder" package for users.

Signed-off-by: Geunsik Lim <geunsik.lim@samsung.com>

leemgs commented:
* Wiki Manual:
![image](https://user-images.githubusercontent.com/82404/65213619-dbb3c800-dae1-11e9-8125-167120b9a1f7.png)

leemgs commented:
* Output after this PR:

![image](https://user-images.githubusercontent.com/82404/65213706-34836080-dae2-11e9-9c6c-bc4068135915.png)

leemgs commented:
To Reviewers, 

This PR required the reviews from the below developers. 
* Mandatory: @jaeyun-jung , @dongju-chae 
* Optional: @wooksong , @myungjoo , Others

dongju-chae commented on ci/taos/plugins-good/pr-format-flawfinder.sh:
```diff
@@ -0,0 +1,131 @@
+        message="Skipped. Your PR does not include c/c++ code(s)."
+        cibot_report $TOKEN "success" "TAOS/pr-format-flawfinder" "$message" "${CISERVER}${PRJ_REPO_UPSTREAM}/ci/${dir_commit}/" "${GITHUB_WEBHOOK_API}/statuses/$input_commit"
+    else
+        echo "[DEBUG] Failed. static code analysis tool for3 security - flawfinder."
```
for3 --> for

leemgs replied:
Nice catch.

leemgs replied:
Fixed.

dongju-chae commented on ci/taos/plugins-good/pr-format-flawfinder.sh:
```diff
@@ -0,0 +1,131 @@
+
+                    flawfinder_result="flawfinder_security_flaw_result"
+                    # Check C/C++ file.
+                    $static_analysis_sw $static_analysis_rules $i > ../report/${flawfinder_result}_${i}.html
```
Does this make a log html file for each source file? Hmm..

leemgs replied:
Good question, Please refer to  Line 98 "**break**". If a file of the commit includes a security flaw, the module reports "failure" status. Then the module does not try to inspect the next file by default as an intended policy.

dongju-chae replied:
Yes. I have two suggestions about this.
1) `break` after inspection of each file: if a commit has some security flaws in multiple files, a user may need to make multiple PRs, which is redundant work.
2) too many html log files: if there is no error, `/report` folder contains too many log files, which means it's difficult for users to check other report logs (maybe, `/report/flawfinder/` folder would be another option).

So, my point it, how about merging these log files into a single report log? and check this merged file using `flawfinder`.

leemgs replied:
> 2. Merge log files: if there is no error, /report folder contains too many log files,
The output structure has a rule consistently. For example, if the source file is nnstreamer-tensor.c, the output is ****_nnstreamer-tensor.txt file. if we try to merge lots of log file into the one file, the readability of the error report will be damaged.

leemgs replied:
> 1. No break after inspection of each file: if a commit has some security flaws in multiple files, a user may need to make multiple force-push to check again and again, which is redundant work.

Right, However developers did not often update multiple log files at once in the past from our experiment result (e.g. CPPcheck module). So, we had changed the report style from "all report" to "step-by-step" style. :) If we meet new cases in the near future, let's customize this module again to give contributors efficiency and convenience.  (And we have been keeping 1pr-1commit rule well.)

leemgs commented:
**Changelog**
* Version 2:
  - Added annotation for the module maintenance.
  - Improved the code readability.

* Version 1:
  - Added new module (e.g. flawfinder) for scanning security flaws.
  - Added configuration varaibles.
  - Added document.
  - Added installation step of "flawfinder" package for users.

jaeyun-jung commented on ci/taos/plugins-good/pr-format-flawfinder.sh:
```diff
@@ -0,0 +1,136 @@
+            # in case of source code files (*.c, *.cpp)
+            case ${i} in
+                # in case of C/C++ code
+                *.c|*.cpp)
```
*.cc

jaeyun-jung replied:
plz update the file extension.

leemgs replied:
Updated.

jaeyun-jung commented on ci/taos/plugins-good/pr-format-flawfinder.sh:
```diff
@@ -0,0 +1,136 @@
+   
+    # Send the webhook message to a GitHub repository.
+    if [[ $check_result == "success" ]]; then
+        echo "[DEBUG] Passed. static code analysis tool for security - flawfinder."
```
Typo : Passed. **S**tatic code
line 117, 121, 125

leemgs replied:
Thanks. Fixed.

leemgs commented:
**Changelog**
* Version 3:
  - Added source code formats (e.g., .cc and .c++)

* Version 2:
  - Added annotation for the module maintenance.
  - Improved the code readability.

* Version 1:
  - Added new module (e.g. flawfinder) for scanning security flaws.
  - Added configuration variables.
  - Added document.
  - Added installation step of "flawfinder" package for users.
